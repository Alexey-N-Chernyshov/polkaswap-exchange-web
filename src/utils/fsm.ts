import { Machine, MachineOptions, assign } from 'xstate'

enum EVENTS {
  SEND_SORA = 'SEND_SORA',
  SEND_ETHEREUM = 'SEND_ETHEREUM',
  RETRY = 'RETRY'
}

enum STATES {
  INITIAL = 'INITIAL',
  SORA_SUBMITTED = 'SORA_SUBMITTED',
  SORA_PENDING = 'SORA_PENDING',
  SORA_REJECTED = 'SORA_REJECTED',
  SORA_COMMITED = 'SORA_COMMITED',
  ETHEREUM_SUBMITTED = 'ETHEREUM_SUBMITTED',
  ETHEREUM_PENDING = 'ETHEREUM_PENDING',
  ETHEREUM_REJECTED = 'ETHEREUM_REJECTED',
  ETHEREUM_COMMITED = 'ETHEREUM_COMMITED'
}

enum ACTIONS {
  SIGN_TRANSACTION = 'SIGN_TRANSACTION',
  SET_TRANSACTION_STATE = 'SET_TRANSACTION_STATE',
  SET_SECOND_TRANSACTION_STEP = 'SET_SECOND_TRANSACTION_STEP',
  SET_SORA_TRANSACTIONS_HASH = 'SET_SORA_TRANSACTIONS_HASH',
  SET_ETHEREUM_TRANSACTION_HASH = 'SET_ETHEREUM_TRANSACTION_HASH'
}

enum SERVICES {
  SIGN_SORA_TRANSACTION = 'SIGN_SORA_TRANSACTION',
  CHECK_SORA_TRANSACTION = 'CHECK_SORA_TRANSACTION',
  SIGN_ETHEREUM_TRANSACTION = 'SIGN_ETHEREUM_TRANSACTION',
  CHECK_ETHEREUM_TRANSACTION = 'CHECK_ETHEREUM_TRANSACTION'
}

const setTransactionState = (state: STATES) => {
  return assign({
    history: (context: any) => {
      console.log('update transaction state', state)
      return ({
        ...context.history,
        transactionState: state
      })
    }
  })
}

const SORA_ETHEREUM_STATES = {
  [STATES.INITIAL]: {
    on: {
      SEND_SORA: STATES.SORA_SUBMITTED
    }
  },
  [STATES.SORA_SUBMITTED]: {
    entry: setTransactionState(STATES.SORA_SUBMITTED),
    invoke: {
      src: SERVICES.SIGN_SORA_TRANSACTION,
      onDone: {
        target: STATES.SORA_PENDING,
        actions: ACTIONS.SIGN_TRANSACTION
      },
      onError: STATES.SORA_REJECTED
    }
  },
  [STATES.SORA_PENDING]: {
    entry: setTransactionState(STATES.SORA_PENDING),
    invoke: {
      src: SERVICES.CHECK_SORA_TRANSACTION,
      onDone: {
        target: STATES.SORA_COMMITED,
        actions: ACTIONS.SET_SORA_TRANSACTIONS_HASH
      },
      onError: STATES.SORA_REJECTED
    }
  },
  [STATES.SORA_REJECTED]: {
    entry: setTransactionState(STATES.SORA_REJECTED),
    on: {
      RETRY: STATES.SORA_SUBMITTED
    }
  },
  [STATES.SORA_COMMITED]: {
    entry: setTransactionState(STATES.SORA_COMMITED),
    on: {
      SEND_ETHEREUM: STATES.ETHEREUM_SUBMITTED
    }
  },
  [STATES.ETHEREUM_SUBMITTED]: {
    entry: [
      setTransactionState(STATES.ETHEREUM_SUBMITTED),
      ACTIONS.SET_SECOND_TRANSACTION_STEP
    ],
    invoke: {
      src: SERVICES.SIGN_ETHEREUM_TRANSACTION,
      onDone: {
        target: STATES.ETHEREUM_PENDING,
        actions: [
          ACTIONS.SIGN_TRANSACTION,
          ACTIONS.SET_ETHEREUM_TRANSACTION_HASH
        ]
      },
      onError: STATES.ETHEREUM_REJECTED
    }
  },
  [STATES.ETHEREUM_PENDING]: {
    entry: setTransactionState(STATES.ETHEREUM_PENDING),
    invoke: {
      src: SERVICES.CHECK_ETHEREUM_TRANSACTION,
      onDone: STATES.ETHEREUM_COMMITED,
      onError: STATES.ETHEREUM_REJECTED
    }
  },
  [STATES.ETHEREUM_REJECTED]: {
    entry: setTransactionState(STATES.ETHEREUM_REJECTED),
    on: {
      RETRY: STATES.ETHEREUM_SUBMITTED
    }
  },
  [STATES.ETHEREUM_COMMITED]: {
    entry: setTransactionState(STATES.ETHEREUM_COMMITED)
  }
}

const ETHEREUM_SORA_STATES = {
  [STATES.INITIAL]: {
    on: {
      SEND_ETHEREUM: STATES.ETHEREUM_SUBMITTED
    }
  },
  [STATES.ETHEREUM_SUBMITTED]: {
    always: {
      target: STATES.ETHEREUM_PENDING
    }
  },
  [STATES.ETHEREUM_PENDING]: {
    invoke: {
      src: SERVICES.CHECK_ETHEREUM_TRANSACTION,
      onDone: STATES.ETHEREUM_COMMITED,
      onError: STATES.ETHEREUM_REJECTED
    }
  },
  [STATES.ETHEREUM_REJECTED]: {
    on: {
      RETRY: STATES.ETHEREUM_PENDING
    }
  },
  [STATES.ETHEREUM_COMMITED]: {
    on: {
      SEND_SORA: STATES.SORA_SUBMITTED
    }
  },
  [STATES.SORA_SUBMITTED]: {
    always: {
      target: STATES.SORA_PENDING
    }
  },
  [STATES.SORA_PENDING]: {
    invoke: {
      src: SERVICES.CHECK_SORA_TRANSACTION,
      onDone: STATES.SORA_COMMITED,
      onError: STATES.SORA_REJECTED
    }
  },
  [STATES.SORA_REJECTED]: {
    on: {
      RETRY: STATES.SORA_SUBMITTED
    }
  },
  [STATES.SORA_COMMITED]: {
    type: 'final'
  }
}
interface TransactionPart {
  sign: () => Promise<any>;
  send: () => Promise<any>;
}
interface TransactionFlow {
  first: TransactionPart;
  second: TransactionPart;
}
interface Actions {
  history: any;
  flow: TransactionFlow;
}

function createFSM (actions: Actions, states, initialState = STATES.INITIAL) {
  type Transitions =
    | { type: EVENTS.SEND_SORA }
    | { type: EVENTS.SEND_ETHEREUM }
    | { type: EVENTS.RETRY }
  interface Context {
    history: any;
    sendSoraTransactionPromise?: ({ txId: string }) => Promise<any>;
    signSoraTransactionPromise?: ({ txId: string }) => Promise<any>;
    sendEthereumTransactionPromise?: ({ ethereumHash: string }) => Promise<any>;
    signEthereumTransactionPromise?: ({ hash: string }) => Promise<any>;
  }

  const transactionActions = states === SORA_ETHEREUM_STATES
    ? [actions.flow.first, actions.flow.second]
    : [actions.flow.second, actions.flow.first]

  console.log(actions.history)

  const initialContext: Context = {
    history: actions.history,
    sendSoraTransactionPromise: transactionActions[0].send,
    signSoraTransactionPromise: transactionActions[0].sign,
    sendEthereumTransactionPromise: transactionActions[1].send,
    signEthereumTransactionPromise: transactionActions[1].sign
  }

  const machineOptions: Partial<MachineOptions<Context, any>> = {
    actions: {
      [ACTIONS.SET_SECOND_TRANSACTION_STEP]: assign({
        history: (context, event) => {
          console.log('second step', 2)
          return ({
            ...context.history,
            transactionStep: 2,
            signed: false
          })
        }
      }),
      [ACTIONS.SIGN_TRANSACTION]: assign({
        history: (context, event) => {
          console.log('sign sora', event)
          return ({
            ...context.history,
            signed: true
          })
        }
      }),
      [ACTIONS.SET_SORA_TRANSACTIONS_HASH]: assign({
        history: (context, event) => {
          console.log('sora hash', event)
          return ({
            ...context.history,
            hash: event.data
          })
        }
      }),
      [ACTIONS.SET_ETHEREUM_TRANSACTION_HASH]: assign({
        history: (context, event) => {
          console.log('ethereum hash', event)
          return ({
            ...context.history,
            ethereumHash: event.data
          })
        }
      })
    },
    services: {
      [SERVICES.SIGN_SORA_TRANSACTION]: (context: Context, event: any): PromiseLike<any> => {
        if (!context.signSoraTransactionPromise) {
          throw new Error('Unexpected behaviour! Please check SORA transaction')
        }

        console.log('SIGN', context.history.id)

        if (context.history.signed) return Promise.resolve()

        return context.signSoraTransactionPromise({
          txId: context.history.id
        })
      },
      [SERVICES.CHECK_SORA_TRANSACTION]: (context: Context, event: any): PromiseLike<any> => {
        if (!context.sendSoraTransactionPromise) {
          throw new Error('Unexpected behaviour! Please check SORA transaction')
        }

        console.log('send', context.history.id)

        return context.sendSoraTransactionPromise({
          txId: context.history.id
        })
      },
      [SERVICES.SIGN_ETHEREUM_TRANSACTION]: (context: Context, event: any): PromiseLike<any> => {
        if (!context.signEthereumTransactionPromise) {
          throw new Error('Unexpected behaviour! Please check ETH transaction')
        }

        return context.signEthereumTransactionPromise({
          hash: context.history.hash
        })
      },
      [SERVICES.CHECK_ETHEREUM_TRANSACTION]: (context: Context, event: any): PromiseLike<any> => {
        if (!context.sendEthereumTransactionPromise) {
          throw new Error('Unexpected behaviour! Please check ETH transaction')
        }

        console.log(context.history.ethereumHash)

        return context.sendEthereumTransactionPromise({
          ethereumHash: context.history.ethereumHash
        })
      }
    }
  }

  return Machine<Context, any, Transitions>(
    {
      initial: initialState,
      context: initialContext,
      states
    },
    machineOptions
  )
}

// INITIAL -- nothing
// *_SUBMITTED - signed transaction
// *_PENDING - sended transaction
// *_REJECTED - rejected transaction, possible to call retry
// *_COMMITED = complete transaction

export {
  createFSM,
  EVENTS,
  STATES,
  SORA_ETHEREUM_STATES,
  ETHEREUM_SORA_STATES
}
