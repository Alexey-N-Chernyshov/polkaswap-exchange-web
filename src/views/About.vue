<template>
  <div class="layout s-flex">
    <div class="background s-flex"></div>
    <div class="content">
      <div class="terms s-flex">
        <span>{{ t('about.polkaswapText') }}</span>
        <div class="web3-logo s-flex">
          <span>{{ t('about.fundedBy') }}</span>
          <i class="logo" />
        </div>
      </div>
      <s-divider />
      <div class="links s-flex">
        <s-button type="primary" size="medium" @click="handleClickExchange">
          {{ t('about.openExchange') }}
        </s-button>
        <s-button type="link" size="medium" icon="external-link" @click="handleGoToMedium">
          {{ t('about.mediumLink') }}
        </s-button>
        <s-button type="link" size="medium" icon="external-link" @click="handleGoToGithub">
          {{ t('about.githubLink') }}
        </s-button>
      </div>
      <s-divider />
      <div class="articles s-flex">
        <s-row>
          <s-col :lg="3" :md="4" :sm="6" :xs="12" v-for="topic in AboutTopics" :key="topic.title">
            <s-card class="article-card" clickable>
              <template #header>
                <div class="article-header">
                  <span class="bold" style="font-size: 18px;">{{ t(`about.${topic.title}.title`) }}</span>
                  <s-icon :size="24" :name="topic.icon" />
                </div>
              </template>
              <span>{{ t(`about.${topic.title}.text`) }}</span>
            </s-card>
          </s-col>
        </s-row>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator'

import TranslationMixin from '@/components/mixins/TranslationMixin'
import { AboutTopics, PageNames } from '@/consts'
import router from '@/router'

@Component
export default class About extends Mixins(TranslationMixin) {
  readonly AboutTopics = AboutTopics

  handleClickExchange (): void {
  }

  handleGoToMedium (): void {
  }

  handleGoToGithub (): void {
  }
}
</script>

<style lang="scss" scoped>
.layout {
  flex-direction: column;
  height: 100%;
  .background {
    min-height: 120px;
    height: 256px;
    background-image: url("~@/assets/background.png");
    background-size: cover;
    background-repeat: no-repeat;
    align-items: center;
    justify-content: center;
  }
  .content {
    padding: 30px 60px;
    .terms {
      > span {
        font-size: 36px;
      }
      .web3-logo {
        flex-direction: column;
        margin-left: 10%;
        > span {
          color: #A1A1A0;
          font-size: 12px;
        }
        .logo {
          margin-top: 12px;
          background-image: url('~@/assets/web3-logo.svg');
          width: 112px;
          height: 48px;
        }
      }
    }
    .articles {
      .article-card {
        background-color: #F5F7F8;
        border-color: transparent;
        border-radius: 24px;
        margin: 4px;
        height: 140px;
        font-size: 14px;
        .article-header {
          display: flex;
          justify-content: space-between;
          i {
            color: var(--s-color-main-brand);
          }
        }
      }
    }
  }
}
@media (max-width: 900px) {
  .layout .content .terms > span {
    font-size: 24px;
  }
}
@media (max-width: 600px) {
  .layout .content .terms > span {
    font-size: 16px;
    font-weight: bold;
  }
}
</style>
