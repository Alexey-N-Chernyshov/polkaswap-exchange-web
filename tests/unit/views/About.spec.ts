import { shallowMount, createLocalVue } from '@vue/test-utils'

import About from '@/views/About.vue'
import { TranslationMock } from '../../utils'

const localVue = createLocalVue()

describe('About.vue', () => {
  beforeEach(() => {
    TranslationMock(About)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(About, { localVue })
    expect(wrapper.element).toMatchSnapshot()
  })
})
