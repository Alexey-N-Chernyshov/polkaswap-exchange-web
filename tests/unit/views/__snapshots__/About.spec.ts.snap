// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About.vue should renders correctly 1`] = `
<div
  class="layout s-flex"
>
  <div
    class="banner s-flex"
  />
   
  <div
    class="content"
  >
    <div
      class="terms s-flex"
    >
      <span>
        
      </span>
       
      <div
        class="web3-logo s-flex"
      >
        <span>
          
        </span>
         
        <i
          class="logo"
        />
      </div>
    </div>
     
    <s-divider />
     
    <div
      class="links s-flex"
    >
      <s-button
        size="medium"
        type="primary"
      >
        
        
      
      </s-button>
       
      <s-button
        icon="external-link"
        size="medium"
        type="link"
      >
        
        
      
      </s-button>
       
      <s-button
        icon="external-link"
        size="medium"
        type="link"
      >
        
        
      
      </s-button>
    </div>
     
    <s-divider />
     
    <div
      class="articles s-flex"
    >
      <s-row>
        <s-col
          lg="3"
          md="4"
          sm="6"
          xs="12"
        >
          <s-card
            class="article-card"
            clickable=""
          >
             
            <span>
              
            </span>
          </s-card>
        </s-col>
        <s-col
          lg="3"
          md="4"
          sm="6"
          xs="12"
        >
          <s-card
            class="article-card"
            clickable=""
          >
             
            <span>
              
            </span>
          </s-card>
        </s-col>
        <s-col
          lg="3"
          md="4"
          sm="6"
          xs="12"
        >
          <s-card
            class="article-card"
            clickable=""
          >
             
            <span>
              
            </span>
          </s-card>
        </s-col>
        <s-col
          lg="3"
          md="4"
          sm="6"
          xs="12"
        >
          <s-card
            class="article-card"
            clickable=""
          >
             
            <span>
              
            </span>
          </s-card>
        </s-col>
      </s-row>
    </div>
  </div>
</div>
`;
